test/auth_aws/test_auth_aws.py

```diff
@@ -53,6 +56,15 @@ def test_connect_uri(self):
         with MongoClient(self.uri) as client:
             client.get_database().test.find_one()
 
+    def test_cache_credentials(self):
+        with MongoClient(self.uri) as client:
+            client.get_database().test.find_one()
+            pool = get_pool(client)
+        pool.reset()
+        with mock.patch.dict(os.environ, {}, clear=True):
+            with MongoClient(self.uri) as client:
+                client.get_database().test.find_one()
```

Comment:
Is this test supposed to fail without caching? If so it would be good to make that explicit by asserting that this fails initially:
```
        # Auth fails without env variables.
        with mock.patch.dict(os.environ, {}, clear=True):
            with MongoClient(self.uri) as client:
                with self.assertRaises(OperationFailure):
                    client.get_database().test.find_one()

        # Auth succeeds and caches credential.
        with MongoClient(self.uri) as client:
            client.get_database().test.find_one()
        # Auth succeeds with cached credential.
        with mock.patch.dict(os.environ, {}, clear=True):
            with MongoClient(self.uri) as client:
                client.get_database().test.find_one()
```

Reply:
This may not be possible to test.  We don't cache when the creds environment variables are given directly, which is why those tests are failing.   We have no way of knowing if cached creds were used on EC2 without using `requests-mock` which seems like too much of invasion into implementation details.

Reply:
We at least need to test what we can. For example, the cache reset on auth failure logic. We can use failCommand to trigger a server side auth failure.

Would it be possible to  use the connectionStatus command to determine which credential was authed?
```
>>> client.admin.command('connectionStatus')
{'authInfo': {'authenticatedUsers': [...]
```
https://www.mongodb.com/docs/manual/reference/command/connectionStatus/#mongodb-data-connectionStatus.authinfo.authenticatedUsers

Probably not I suppose, since I imagine different temp creds will map to the same database level user.

Reply:
Or a test that "poisons" the cache with invalid creds. The first auth attempt should fail and clear the cache, the next attempt should generate a new cred and succeed.